### PR TITLE
Fix #2826

### LINKED ISSUE #2826: Deprecation warning in plugins.py
### Problem

Installed beets from fedora repo and getting the following error every time I run any command

Running this command in verbose (`-vv`) mode:

```sh
[akya@quady scripts] $ beet -vv
user configuration: /home/akya/.config/beets/config.yaml
data directory: /home/akya/.config/beets
plugin paths: /home/akya/.config/beets/beetsplug
/usr/lib/python3.6/site-packages/urllib3/contrib/pyopenssl.py:46: DeprecationWarning: OpenSSL.rand is deprecated - you should use os.urandom instead
  import OpenSSL.SSL
Sending event: pluginload
/usr/lib/python3.6/site-packages/beets/plugins.py:114: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() or inspect.getfullargspec()
  argspec = inspect.getargspec(func)
lyrics: Disabling google source: no API key configured.
library database: /home/akya/data/musiclibrary.db
library directory: /media/Media/Songs/English
Sending event: library_opened
Usage: 
  beet COMMAND [ARGS...]
  beet help COMMAND

Options:
  --format-item=FORMAT_ITEM
                        print with custom format
  --format-album=FORMAT_ALBUM
                        print with custom format
  -l LIBRARY, --library=LIBRARY
                        library database file to use
  -d DIRECTORY, --directory=DIRECTORY
                        destination music directory
  -v, --verbose         log more details (use twice for even more)
  -c CONFIG, --config=CONFIG
                        path to configuration file
  -h, --help            show this help message and exit

Commands:
  bad               check for corrupt or missing files
  clearart          remove images from file metadata
  config            show or edit the user configuration
  convert           convert to external location
  duplicates (dup)  List duplicate tracks or albums.
  embedart          embed image files into file metadata
  extractart        extract an image from file metadata
  fetchart          download album art
  fields            show fields available for queries and format strings
  help (?)          give detailed help on a specific sub-command
  import (imp, im)  import new music
  info              show file metadata
  lastgenre         fetch genres
  list (ls)         query the library
  lyrics            fetch song lyrics
  missing (miss)    List missing tracks.
  modify (mod)      change metadata fields
  move (mv)         move or copy items
  play              send music to a player as a playlist
  remove (rm)       remove matching items from the library
  splupdate         update the smart playlists. Playlist names may be passed
                    as arguments.
  stats             show statistics about the library or a query
  update (upd, up)  update the library
  version           output version information
  web               start a Web interface
  write             write tag information to files
Sending event: cli_exit

```

Led to this problem:

```
[akya@quady scripts] $ beet --version
/usr/lib/python3.6/site-packages/urllib3/contrib/pyopenssl.py:46: DeprecationWarning: OpenSSL.rand is deprecated - you should use os.urandom instead
  import OpenSSL.SSL
/usr/lib/python3.6/site-packages/beets/plugins.py:114: DeprecationWarning: inspect.getargspec() is deprecated, use inspect.signature() or inspect.getfullargspec()
  argspec = inspect.getargspec(func)
beets version 1.4.6
Python version 3.6.4
plugins: badfiles, beatport, convert, duplicates, embedart, fetchart, info, lastgenre, lyrics, missing, mpdupdate, play, smartplaylist, web

```

Here's a link to the music files that trigger the bug (if relevant): 

### Setup

* OS: Fedora 27
* Python version: 3.6.4 
* beets version: 1.4.6
* Turning off plugins made problem go away (yes/no): yes, problem is in plugins.py

My configuration (output of `beet config`) is:

```yaml
directory: /media/Media/Songs/English
#directory: /media/Media/Songs/Hindi
library: ~/data/musiclibrary.db
#library: ~/data/musiclibrary2.db
import:
    copy: no
    write: yes
    log: /home/akya/.config/beets/import.log

pluginpath: 
    - /home/akya/.config/beets/beetsplug/

plugins: fetchart embedart lastgenre lyrics web missing beatport duplicates convert info play smartplaylist mpdupdate badfiles

play:
    command: mpv $args --playlist

mpd:
    host: 192.168.1.100
    port: 6600

fetchart:
    cautious: true
    cover_names: front back
    minwidth: 300
    sources: itunes *

smartplaylist:
    relative_to: 
    playlist_dir: ~/.mpd/playlists
    playlists:
        - name: fav.m3u
          query: 'comments:Fav'
        - name: new.m3u
          query: 'comments:New'
        - name: epic.m3u
          query: 'comments:Epic'
        - name: lush.m3u
          query: 'comments:Lush'
        - name: naruto.m3u
          query: 'comments:Naruto'
        - name: latest.m3u
          query: 'added:2017-11-27'
convert:
    format: mp3
    dest: /home/akya/1Dump/sync
    max_bitrate: 320
    never_convert_lossy_files: yes


paths:
    default: $albumartist/$album%aunique{}/$track $title
#    default: $album%aunique{}/$track $title
    singleton: Non-Album/$artist - $title
    comp: Compilations/$album%aunique{}/$track $title

```

### PR DESCRIPTION
**Pull request by zhelezov**
Test for major Python version and use inspect.getargspec() or
inspect.getfullargspec() respectively to silence deprecation warnings in
Python 3 as discussed in #2826

### CODE REVIEW

**Comment by zhelezov** [comment]
Having another look at this I don't like the fact that `argspec` is of a different type in the two branches: 4-tuple of type `ArgSpec(args, varargs, keywords, defaults)` in the PY2 case, and 8-tuple of type `FullArgSpec(args, varargs, varkw, defaults, kwonlyargs, kwonlydefaults, annotations)` in the PY3 branch. The above proposed simple solution is alright because we only ever use the first member `argspec.args` but it is a possible source of future confusion should one try to use the third member which is not accessible by the same attribute: `argspec.keywords` vs `argspec.varkw`.

To make the tuple identical in both cases I propose two solutions.

The simple one:
```diff
diff --git a/beets/plugins.py b/beets/plugins.py
index 69784d26..9a984cf5 100644
--- a/beets/plugins.py
+++ b/beets/plugins.py
@@ -130,7 +130,7 @@ class BeetsPlugin(object):
         if six.PY2:
             argspec = inspect.getargspec(func)
         else:
-            argspec = inspect.getfullargspec(func)
+            argspec = inspect.ArgSpec(*inspect.getfullargspec(func)[:4])

         @wraps(func)
         def wrapper(*args, **kwargs):
```
What bugs me about this is that if `inspect.getargspec()` is removed from Python 3.8, so would most likely be `inspect.ArgSpec` and this will soon break.

The second solution is two replicate `inspect.ArgSpec` explicitly, sth like this:
```diff
diff --git a/beets/plugins.py b/beets/plugins.py
index 69784d26..12d9dde5 100644
--- a/beets/plugins.py
+++ b/beets/plugins.py
@@ -21,6 +21,7 @@ import inspect
 import traceback
 import re
 from collections import defaultdict
+from collections import namedtuple
 from functools import wraps


@@ -130,7 +131,9 @@ class BeetsPlugin(object):
         if six.PY2:
             argspec = inspect.getargspec(func)
         else:
-            argspec = inspect.getfullargspec(func)
+            ArgSpec = namedtuple('ArgSpec', 'args varargs keywords defaults',
+                                 defaults=([], None, None, ()))
+            argspec = ArgSpec(*inspect.getfullargspec(func)[:4])

         @wraps(func)
         def wrapper(*args, **kwargs):
```

What are your thoughts about my concerns? Leave it like it is, maybe just add a comment with a warning, or commit the second proposed solution which looks more robust, if a bit clunkier, to me?

**Comment by zhelezov** [comment]
A third variant, stolen more or less from Django's code base, is to use `inspect.signature` which is the recommended aproach in Python 3. There was a proposal from Python devs to get rid from both `getargspec` and `getfullargspec` even if for the moment they seem to have backed off from the decision to deprecate the latter. In that case we can introduce the wrapper in say `beets/util/inspect.py`:
```python
from __future__ import absolute_import

import inspect
from collections import namedtuple

from six import PY2


def getargspec(func):
    if PY2:
        return inspect.getargspec(func)

    ArgSpec = namedtuple('ArgSpec', 'args varargs keywords defaults')

    sig = inspect.signature(func)
    args = [
        p.name for p in sig.parameters.values()
        if p.kind == inspect.Parameter.POSITIONAL_OR_KEYWORD
    ]
    varargs = [
        p.name for p in sig.parameters.values()
        if p.kind == inspect.Parameter.VAR_POSITIONAL
    ]
    varargs = varargs[0] if varargs else None
    varkw = [
        p.name for p in sig.parameters.values()
        if p.kind == inspect.Parameter.VAR_KEYWORD
    ]
    varkw = varkw[0] if varkw else None
    defaults = tuple(p.default for p in sig.parameters.values()
                     if p.kind == inspect.Parameter.POSITIONAL_OR_KEYWORD
                     and p.default is not p.empty) or None

    return ArgSpec(args, varargs, varkw, defaults)
```

And import it elsewhere in place of `inspect` from stdlib. In the case of `beets/plugins.py` the sole change needed would be the import:
```diff
diff --git a/beets/plugins.py b/beets/plugins.py
index 6dec7ef2..56bb07cb 100644
--- a/beets/plugins.py
+++ b/beets/plugins.py
@@ -17,7 +17,6 @@

 from __future__ import division, absolute_import, print_function

-import inspect
 import traceback
 import re
 from collections import defaultdict
@@ -27,6 +26,7 @@ from functools import wraps
 import beets
 from beets import logging
 from beets import mediafile
+from beets.util import inspect
 import six

 PLUGIN_NAMESPACE = 'beetsplug'
```

**Comment by bb010g** [comment]
The wrapper option looks like the best path forwards to me.

**Comment by ssssam** [comment]
I've tested this PR and it fixes the deprecation warnings issue for me. Shall we merge it? Perhaps with a link back to the comment above documenting the potential gotcha for future devs?

**Comment by zhelezov** [comment]
I am sorry about the delay, almost forgot about this issue actually. I pushed the wrapper option from my last post as suggested. Feel free to do as you please. Cheers.

**Comment by sampsyo** [comment]
This does look like quite a complete solution! The only thing is: when we (hopefully before too long) transition to Python 3 exclusively, I predict that we may not need perfect emulation of `getargspec`. We only need to pull out a few attributes for our AST-processing purposes, which makes this (very clean!) drop-in replacement somewhat overkill. But it's probably fine for now.

**Comment by zhelezov** [comment]
Thought about the same thing. That's why I isolated the workaround in a separate file which can simply be removed after the transition of the code base to Python 3. I don't see a cleaner solution having in mind backward compatibility.

**Comment by sampsyo** [comment]
Indeed; seems like a great idea for this to get isolated in its own separate file. The thing I was bringing up, however, is that we're currently emulating a Python 2 API on Python 3, not the other way around. So when we make the transition, we won't be able to just remove the file---we'll be able to remove the `if PY2:` early abort, but more refactoring will be required to use the new API "directly."

File: beets/util/inspect.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+from collections import namedtuple
+
+from six import PY2
+
+
+def getargspec(func):
+    if PY2:
+        return inspect.getargspec(func)
+
+    ArgSpec = namedtuple('ArgSpec', 'args varargs keywords defaults')
```
Comment (sampsyo) [inline beets/util/inspect.py]: It’s customary to put namedtuple declarations at the top (module) level.

Comment (zhelezov) [reply beets/util/inspect.py]: Ok, I'll move it out to the top level.

**Comment by sampsyo** [comment]
Fantastic; thank you!

**Pull request merged**